add PAG support for Stable Diffusion 3

sunovivid opened the pull request:
# What does this PR do?

This is a  PR for adding PAG support for SD3!

SD3 differs slightly from SD and SDXL because it employs Rectified Flow and uses the MMDiT backbone, not UNet.

For the joint attention in MMDiT, we can apply perturbed self-attention by masking attention between image patches, following the principles of PAG.

It works quite well. Here is an example with PAG + SD3.

# Examples
_"Pirate ship trapped in a cosmic maelstrom nebula"_
![resized_pag](https://github.com/user-attachments/assets/0bc6f6ae-89dc-4546-a143-7803df587c6d)

From left to right, the PAG scale increases to 0.5, 1.0, 3.0, 5.0, and 7.0.
From top to bottom, the CFG scale increases to 1.0, 3.0, 5.0, and 7.0.

# How to Use
```
import torch
from diffusers import AutoPipelineForText2Image

pipe = AutoPipelineForText2Image.from_pretrained("stabilityai/stable-diffusion-3-medium-diffusers", torch_dtype=torch.float16, enable_pag=True, pag_applied_layers=["13"])
pipe.to("cuda")

prompt = "a photo of a cat holding a sign that says hello world"

image = pipe(
    prompt=prompt,
    negative_prompt="",
    num_inference_steps=28,
    height=1024,
    width=1024,
    guidance_scale=0,
    pag_scale=5.0,
).images[0]

image.save(f"result.png")
```

# Ablations on Layer
For those interested in knowing which layer to perturb, I am providing the results of single-layer perturbation for all the MMDiT attention layers in SD3.

<details>
<summary>ablations</summary>
from left to right, guidance scale: 3.0, 5.0, 7.0

from top to bottom, `pag_applied_layer`: 0,1,2,3
![sd3_layer_a photo of a ca_seed_0_slice_1](https://github.com/user-attachments/assets/2074f383-a3f5-4cba-962a-f7bcb0dfaf2b)
from top to bottom, `pag_applied_layer`: 4,5,6,7
![sd3_layer_a photo of a ca_seed_0_slice_2](https://github.com/user-attachments/assets/d42ecb7c-7e38-49cd-921b-0142cb1f76b1)
from top to bottom, `pag_applied_layer`: 8,9,10,11
![sd3_layer_a photo of a ca_seed_0_slice_3](https://github.com/user-attachments/assets/12854606-0184-41d2-95de-77511589df43)
from top to bottom, `pag_applied_layer`: 12,13,14,15
![sd3_layer_a photo of a ca_seed_0_slice_4](https://github.com/user-attachments/assets/b78fa3a0-7885-4f1c-b1dd-2ee9af76be2d)
from top to bottom, `pag_applied_layer`: 16,17,18,19
![sd3_layer_a photo of a ca_seed_0_slice_5](https://github.com/user-attachments/assets/d720a34c-cb55-4b88-926a-92bfa383deba)
from top to bottom, `pag_applied_layer`: 20,21,22,23
![sd3_layer_a photo of a ca_seed_0_slice_6](https://github.com/user-attachments/assets/2ca8bcc3-8275-4617-ab11-7d91353679d1)

</details>


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?
@yiyixuxu @asomoza

@sayakpaul @a-r-r-o-w also could be interested

Thank you for your time!

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @yiyixuxu
- Pipelines and pipeline callbacks: @yiyixuxu and @asomoza
- Training examples: @sayakpaul
- Docs: @stevhliu and @sayakpaul
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @sayakpaul @yiyixuxu @DN6

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

sayakpaul commented on src/diffusers/models/attention_processor.py:
```diff
@@ -1074,6 +1074,300 @@ def __call__(
+
+    def __init__(self):
+        if not hasattr(F, "scaled_dot_product_attention"):
+            raise ImportError("AttnProcessor2_0 requires PyTorch 2.0, to use it, please upgrade PyTorch to 2.0.")
```
```suggestion
            raise ImportError("PAGJointAttnProcessor2_0 requires PyTorch 2.0, to use it, please upgrade PyTorch to 2.0.")
```

sayakpaul commented on src/diffusers/models/attention_processor.py:
```diff
@@ -1074,6 +1074,300 @@ def __call__(
+        batch_size = encoder_hidden_states_ptb.shape[0]
+
+        # `sample` projections.
+        query = attn.to_q(hidden_states_ptb)
```
Maybe let's add a `ptb` suffix to these projection outputs?

sayakpaul commented on src/diffusers/models/attention_processor.py:
```diff
@@ -1074,6 +1074,300 @@ def __call__(
+        value = attn.to_v(hidden_states_ptb)
+
+        # `context` projections.
+        encoder_hidden_states_org_query_proj = attn.add_q_proj(encoder_hidden_states_ptb)
```
Same as above.

sayakpaul commented on src/diffusers/models/attention_processor.py:
```diff
@@ -1074,6 +1074,300 @@ def __call__(
+        seq_len = query.size(2)
+        full_mask = torch.zeros((seq_len, seq_len), device=query.device, dtype=query.dtype)
+
+        # set the attention value between image patches to -inf
+        full_mask[:identity_block_size, :identity_block_size] = float('-inf')
+
+        # set the diagonal of the attention value between image patches to 0
+        full_mask[:identity_block_size, :identity_block_size].fill_diagonal_(0)
```
Do you have any intuition as why do we mask out the image patches in MMDiT while applying identity matrix to the image projections in SDXL?

sayakpaul commented on src/diffusers/models/attention_processor.py:
```diff
@@ -1074,6 +1074,300 @@ def __call__(
+
+    def __init__(self):
+        if not hasattr(F, "scaled_dot_product_attention"):
+            raise ImportError("AttnProcessor2_0 requires PyTorch 2.0, to use it, please upgrade PyTorch to 2.0.")
```
```suggestion
            raise ImportError("PAGCFGJointAttnProcessor2_0 requires PyTorch 2.0, to use it, please upgrade PyTorch to 2.0.")
```

sayakpaul reviewed (approved):
I have some minor comments but other than those this looks quite sleek!

asomoza commented:
I gave this a test, it does fix a little of the generations but I find that the improvement is not as big and noticeable as with SDXL. I still think it's a good addition and probably will help with the controlnets. 

|W/O PAG|PAG|PAG|
|---|---|---|
|![20240714191606_2986130010](https://github.com/user-attachments/assets/5c42d309-4ce2-4cf2-a168-9c44342a2b5a)|![20240714192948_2986130010_6_11](https://github.com/user-attachments/assets/e814ef33-59fb-4953-94fa-33b24c5f84d4)|![20240714193013_2986130010_2_9](https://github.com/user-attachments/assets/40ae070d-b0ed-4839-910a-05174590bd50)|

The pag results are with different layers, but if I use more than 2 layers the generation looks more fake and cartoonish. Also I had to lower the scale to 2.5 or it alos makes it too cartoonish

With both PAG results we can see that the "little people" have better details, but overall I like the style of the original, so probably I will just inpaint and make better the first generation. 

If I use a more realistic photo, the fake and cartoon effect it's even more noticeable, contrary to SDXL that made them more realistic and well defined.

yiyixuxu commented on src/diffusers/pipelines/pag/pag_utils.py:
```diff
@@ -216,6 +281,19 @@ def set_pag_applied_layers(self, pag_applied_layers):
 
         self.pag_applied_layers = pag_applied_layers
+        
+    def set_pag_applied_layers_sd3(self, pag_applied_layers):
```
let's maybe make a  `pag_sd3` folder with a new`SD3PAGMixin` with a `set_pag_applied_layers`, `_set_pag_attn_processor` and `_check_input_pag_applied_layer` specific to sd3

cc @DN6 here let me know what you think!

yiyixuxu replied:
so we just just one "pag" folder then? ok with that

yiyixuxu reviewed (commented):
thanks a ton for working on this!

asomoza commented:
I was playing again with this because I needed a cool generation and  I gave it a revisit. I found a cool application that I'd like to share.

What I found are two things, I was playing with a pag scale that was too high, just using something like 0.5 or 0.7 gives good results that clean a little the image, for example:

|original|0.5|0.7|
|---|---|---|
|![20240719091424_4114929549](https://github.com/user-attachments/assets/ccf72a46-e022-41df-bb0b-f993af5477cb)|![20240719092845_4114929549](https://github.com/user-attachments/assets/6c2aeb56-d28c-43b5-9538-7f265fa27340)|![20240719092932_4114929549](https://github.com/user-attachments/assets/b5176722-3c8f-47fd-bfc8-fe3e2516b259)|

The original has a lot of details which always makes me impressed on how good the SD3 VAE is, but it has some defects like the shape of the tires, the weirdness of the front bumper among other things

Just applying a low pag scale with specific layers, cleans the image a lot and makes the generation better, the only problem is that it makes the truck cleaner which is wrong, but maybe that can be fixed with some other layers or just use this as base for a better generation.

The second thing I found is that we can change the "style" of the generation with just applying pag, for example, without doing anything in the prompt I can change this generation to a more concept art style which I think its really cool.

|original|with pag|
|---|---|
|![20240719091424_4114929549](https://github.com/user-attachments/assets/15e55c91-4e65-47a6-8593-1b8c9dacb486)|![20240719093713_4114929549](https://github.com/user-attachments/assets/e4358809-36d8-4139-a8e7-92598c5a97c4)|

And if I change the prompt to "concept art" instead of "photo":

|original|with pag|
|---|---|
|![20240719100523_1765814182](https://github.com/user-attachments/assets/3440b4c9-1b9f-4bed-b2ce-13da70727f45)|![20240719100413_4114929549](https://github.com/user-attachments/assets/354e9940-575a-4158-b048-406d2a2c33c2)|

It seems to help to keep the composition consistent even if you change the style of the prompt.

sayakpaul commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,934 @@
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
```
To be changed I guess?

sayakpaul commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,936 @@
+
+        if self.do_perturbed_attention_guidance:
+            original_attn_proc = self.transformer.attn_processors
+            self._set_pag_attn_processor(
```
Should this not be `_set_pag_attn_processor_sd3`?

sunovivid commented:
@yiyixuxu @DN6 Thank you for your hard workings with PAG! 
I want to merge this PR, but it seems to be stalled. What steps remain for this PR? I believe I have revised the code following all the comments. Any suggestions?

a-r-r-o-w commented:
@sunovivid Just for notification - we merged #8936 now. It should be easier to now directly apply PAG without the extra SD3PAG class. You'll have to pass PAG related SD3 attention processors as a parameter when calling `self.set_pag_applied_layers(..., pag_attn_processors=(...))` (you can look at how it's done in Hunyuan for an example)

Thanks for the amazing work here! Although a bit unrealistic, it might be nice to get this ready by tomorrow since a Diffusers release will be happening soon and it'd be nice to ship a PAG variant for SD3.

sunovivid commented:
Ok! I want to get a chance. I will try it now. Thank you for your awesome refactoring!

sunovivid commented:
@a-r-r-o-w It seems to be working now. I have integrated your implementation, and the previous implementation still works as before. I also added documentation. However, I was unable to add tests due to a lack of relevant knowledge (e.g., internal model, etc.).

Here’s what I confirmed:

Layer ablation works, and I verified that the results from layers 8-12 are the same as the layer ablation results attached at the beginning of this PR.
Using the prompt "Pirate ship trapped in a cosmic maelstrom nebula," I created grids based on CFG and PAG scale and found similar results to the grids attached at the beginning of this PR, confirming the expected effects according to CFG and PAG (exact seed not recorded, so not reproducible).

Please let me know if there is anything more needed for merging! Your brilliant implementation made the integration super simple. Thank you!

sunovivid commented:
@a-r-r-o-w @yiyixuxu I added SD3 PAG pipelines for `__init__.py`. Could you approve test workflows?

a-r-r-o-w commented on src/diffusers/models/attention_processor.py:
```diff
@@ -1106,6 +1106,326 @@ def __call__(
+            batch_size, channel, height, width = encoder_hidden_states.shape
+            encoder_hidden_states = encoder_hidden_states.view(batch_size, channel, height * width).transpose(1, 2)
+
+        identity_block_size = hidden_states.shape[
```
minor nit: could we maybe move the comment above the statement here so that `make style` does not complain and make this change? not really a fan of the index being on a separate line

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,947 @@
+        ...     "stabilityai/stable-diffusion-3-medium-diffusers",
+        ...     torch_dtype=torch.float16,
+        ...     enable_pag=True,
+        ...     pag_applied_layers=["13"],
```
```suggestion
        ...     pag_applied_layers=["blocks.13"],
```

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,947 @@
+            pag_applied_layers, pag_attn_processors=(PAGCFGJointAttnProcessor2_0(), PAGJointAttnProcessor2_0())
+        )
+
+    def _get_t5_prompt_embeds(
```
Could you add a `# Copied from` statement here from the SD3 file?

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,947 @@
+
+        return prompt_embeds
+
+    def _get_clip_prompt_embeds(
```
`# Copied from` comment here too

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,947 @@
+
+        return prompt_embeds, pooled_prompt_embeds
+
+    def encode_prompt(
```
Not sure if this is exactly the same as in SD3, but if so, `# Copied from` here too

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,947 @@
+
+        return prompt_embeds, negative_prompt_embeds, pooled_prompt_embeds, negative_pooled_prompt_embeds
+
+    def check_inputs(
```
`# Copied from` here too if required

a-r-r-o-w commented on src/diffusers/pipelines/pag/pipeline_pag_sd_3.py:
```diff
@@ -0,0 +1,947 @@
+        if max_sequence_length is not None and max_sequence_length > 512:
+            raise ValueError(f"`max_sequence_length` cannot be greater than 512 but is {max_sequence_length}")
+
+    def prepare_latents(
```
`# Copied from` here too if required

a-r-r-o-w reviewed (commented):
@sunovivid Thank you for the kind words! This is really close to merging, just a couple of things:
- Could you run `make fix-copies`?
- Could you add a test for SD3? It mostly should be a copy of the [SD3 tests](https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion_3/test_pipeline_stable_diffusion_3.py) with two additional tests similar to [this](https://github.com/huggingface/diffusers/blob/a57a7af45cbef004c38e2a294a6457f7f3574e5d/tests/pipelines/pag/test_pag_hunyuan_dit.py#L284) and [this](https://github.com/huggingface/diffusers/blob/a57a7af45cbef004c38e2a294a6457f7f3574e5d/tests/pipelines/pag/test_pag_hunyuan_dit.py#L323)

sunovivid commented:
@a-r-r-o-w I fixed the code following all your comments and added tests! Thank you for your comments and guidance.
I copied only fast tests of SD3 test following @sayakpaul's comments for now. WDYT?

a-r-r-o-w commented on src/diffusers/pipelines/__init__.py:
```diff
@@ -540,6 +541,7 @@
             AnimateDiffPAGPipeline,
             HunyuanDiTPAGPipeline,
             PixArtSigmaPAGPipeline,
+            StableDiffusion3Pipeline,
```
```suggestion
            StableDiffusion3PAGPipeline,
```

to fix the last broken test

a-r-r-o-w commented:
Also, another test seems to be failing: 
![image](https://github.com/user-attachments/assets/0bfe1db3-22bc-4f54-835f-84e7a7daa9f6)

This is probably because `num_layers=1` but we are trying to find the second block using `blocks.1`. You can remove `blocks.1` from that specific failing test to fix I think

sunovivid commented:
I see! How about set `num_layers=2` as in the HunyuanDiTPAG test? I think it is the more robust way (we can test regex-based layer matching)
![image](https://github.com/user-attachments/assets/135ab3c1-c6e7-4969-91e2-2a35538965b8)

a-r-r-o-w commented:
That's sounds good as well. I tried it and there isn't a significant impact on speed (a few extra seconds but that's okay and should be good to test with atleast 2 layers)

sunovivid commented:
Thanks! I set `num_layes=2`.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_8861). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sunovivid commented:
OK. There is another error. Sorry for bothering you. I'll check it out.

sunovivid commented:
@a-r-r-o-w  Thank you for your patience! The failure in the first test case was due to using joint attention in SD3, which uses the name `attn` without distinguishing between `attn1` and `attn2`. I have fixed this error.

The second test case failure was harder to pinpoint. It occurred because, as this is a randomly initialized model, the random weights caused it to turn into `nan` at a specific point (specifically, when `hidden_state_ptb` passed through `attn.to_out`). Thus, the assertion `assert np.abs(out.flatten() - out_pag_enabled.flatten()).max() > 1e-3` failed as it tried to compare `nan` with `0.001`. I believe this is due to the randomness, and when actually sampling results, it works correctly.

<details>
<summary>The part where `nan` occurs:</summary>

![image](https://github.com/user-attachments/assets/18221668-b2d4-4e6b-9c74-67123d1d08c1)
![image](https://github.com/user-attachments/assets/9dc85a06-06e4-460b-9035-ddddffd82e79)
</details>

Sample grid  using CFG and PAG (from left to right, PAG=[1.0, 1.5, 3.0, 5.0], from top to bottom, CFG=[0.5, 0.75, 1.0, 3.0]):

_a photo of a cat holding a sign that says hello world_
![gpu0_prompt_a_photo_resized](https://github.com/user-attachments/assets/a5cdd7ae-f403-4374-a942-0e1c8123f8ad)


_19th century Scottish wizard with a mysterious smile and a piercing gaze, enigmatic, photorealistic, incredibly detailed, sharpness, detail, cinematic lighting_
![gpu1_prompt_19th_century_resized](https://github.com/user-attachments/assets/53acb329-3be6-4c36-9933-e09e651a904b)

<details>

<summary>Sampling code I used (a bit dirty bur for reproducibility):</summary>

```import torch
from diffusers import AutoPipelineForText2Image
from diffusers.utils import load_image, make_image_grid
import multiprocessing

gpu_list = [0, 1]

def sample(gpu_id, prompts):
    pipe = AutoPipelineForText2Image.from_pretrained(
        "stabilityai/stable-diffusion-3-medium-diffusers",
        torch_dtype=torch.float16,
        enable_pag=True, pag_applied_layers=["blocks.9"]
    )
    pipe.to(f"cuda:{gpu_id}")

    seeds = range(1, 6)
    pag_scale_list = [0.5, 0.75, 1.0, 3.0,]
    cfg_scale_list = [1.0, 1.5, 3.0, 5.0,]

    for prompt in prompts:
        for seed in seeds:
            images = []
            for cfg_scale in cfg_scale_list:
                for pag_scale in pag_scale_list:
                    generator = torch.Generator(device="cpu").manual_seed(seed)
                    print(f'GPU {gpu_id} - Prompt: {prompt[:15]} - Seed: {seed} - CFG: {cfg_scale} - PAG: {pag_scale}')
                    image = pipe(
                        prompt=prompt,
                        negative_prompt="",
                        num_inference_steps=28,
                        height=1024,
                        width=1024,
                        guidance_scale=cfg_scale,
                        pag_scale=pag_scale,
                        generator=generator,
                    ).images[0]
                    images.append(image)

            grid_image = make_image_grid(images, rows=len(cfg_scale_list), cols=len(pag_scale_list))
            grid_image.save(f"gpu{gpu_id}_prompt:{prompt[:15]}_seed:{seed}.png")

if __name__ == "__main__":
    prompts = [
        "a photo of a cat holding a sign that says hello world",
        "a photo of an astronaut riding a horse on mars",
        "A stylish woman walks down a Tokyo street filled with warm glowing neon and animated city signage. She wears a black leather jacket, a long red dress, and black boots, and carries a black purse. She wears sunglasses and red lipstick. She walks confidently and casually. The street is damp and reflective, creating a mirror effect of the colorful lights. Many pedestrians walk about.",
        "A close-up photo of a person. The subject is a woman. She wore a blue coat with a gray dress underneath. She has blue eyes and blond hair, and wears a pair of earrings. Behind are blurred city buildings and streets.",
        "two white cats playing on top of the orange sofa, very comfortable Best quality",
        "an illustration of a stylish swordsman",
        "19th century Scottish wizard with a mysterious smile and a piercing gaze, enigmatic, photorealistic, incredibly detailed, sharpness, detail, cinematic lighting",
        "3D animation of a small, round, fluffy creature with big, expressive eyes explores a vibrant, enchanted forest. The creature, a whimsical blend of a rabbit and a squirrel, has soft blue fur and a bushy, striped tail. It hops along a sparkling stream, its eyes wide with wonder. Flowers that glow and change colors, trees with leaves in shades of purple and silver, and small floating lights that resemble fireflies. The creature stops to interact playfully with a group of tiny, fairy-like beings dancing around a mushroom ring.",
        "Pirate ship trapped in a cosmic maelstrom nebula",
        "Oppenheimer sits on the beach on a chair, watching a nuclear exposition with a huge mushroom cloud, 120mm",
    ]

    # Distribute prompts among GPUs
    num_gpus = len(gpu_list)
    prompts_per_gpu = len(prompts) // num_gpus
    distributed_prompts = [prompts[i * prompts_per_gpu:(i + 1) * prompts_per_gpu] for i in range(num_gpus)]

    # Handle any remaining prompts
    remaining_prompts = prompts[num_gpus * prompts_per_gpu:]
    for i in range(len(remaining_prompts)):
        distributed_prompts[i % num_gpus].append(remaining_prompts[i])

    processes = []
    for gpu_id, gpu_prompts in zip(gpu_list, distributed_prompts):
        p = multiprocessing.Process(target=sample, args=(gpu_id, gpu_prompts))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```


</details>


Therefore, I have removed the assert statement in the test. I think it is sufficient if `out_pag_enabled = pipe_pag(**inputs).images[0, -3:, -3:, -1]` generates samples without errors, and the results of `pag_scale=3.0` and `pag_scale=0.0` are not always guaranteed to be different. If you want to include it, you can test it using `nan_to_num` like the following, but I don't think it's necessary:
`python
assert np.max(np.abs(np.nan_to_num(out, nan=0) - np.nan_to_num(out_pag_enabled, nan=0)).flatten()) > 1e-3
`


A new test case failure has been added, which I am not aware of. It seems to be related to the internal model. Do you know why this failure case was added?

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/schedulers/test_schedulers.py::SchedulerBaseTests::test_default_arguments_not_in_config - FileNotFoundError: [Errno 2] No such file or directory: '/github/home/.cache/huggingface/hub/models--hf-internal-testing--tiny-stable-diffusion-pipe/snapshots/3ee6c9f225f088ad5d35b624b6514b091e6a4849/unet/diffusion_pytorch_model.bin'
===== 1 failed, 1445 passed, 406 skipped, 82 warnings in 60.15s (0:01:00) ======
```
---

sayakpaul commented:
Thanks for your awesome contributions, @sunovivid!
